[Mempool] Fix mempool limiting and replace-by-fee for PrioritiseTransaction

sdaftuar opened the pull request:
This fixes the behavior of the mempool limiting to respect any fee deltas applied via PrioritiseTransaction.  

~~(Note that the first 5 commits here are introduced by @morcos in #6898.)~~

It also includes a couple other fixes relating to PrioritiseTransaction:
- Updates RBF logic
- Updates mempool acceptance logic

sdaftuar committed:
Fix mempool limiting for PrioritiseTransaction

Redo the feerate index to be based on mining score, rather than fee.

Update mempool_packages.py to test prioritisetransaction's effect on
package scores.

sdaftuar commented:
This will create merge conflicts for #6871 -- let's try to get that reviewed and merged first, and I can resolve the merge conflicts in this PR afterward.

petertodd commented:
Note that this leaks information about what transactions you've prioritised to the outside world even more so than before. Probably unavoidable right now, but in the long run it'd be better if we had a separate mempool for mining that kept priority deltas. Transactions themselves can be shared across both mempools and reference counted.

petertodd commented on src/rpcblockchain.cpp:
```diff
@@ -244,13 +245,14 @@ UniValue getrawmempool(const UniValue& params, bool fHelp)
             "    \"descendantcount\" : n,  (numeric) number of in-mempool descendant transactions (including this one)\n"
             "    \"descendantsize\" : n,   (numeric) size of in-mempool descendants (including this one)\n"
-            "    \"descendantfees\" : n,   (numeric) fees of in-mempool descendants (including this one)\n"
+            "    \"descendantfees\" : n,   (numeric) modified fees of in-mempool descendants (including this one)\n"
```
The term "modified fees" won't mean anything to the reader. How about "fees of in-mempool descendants (including this one) with priority deltas applied"

sdaftuar replied:
I was trying to make a reference to the term "modifedfee" which I added above without having to redefine the concept, what if I just do: "modified fees (see above) of in-mempool descendants (including this one)"

petertodd commented:
Otherwise, utACK, though will want to recheck rebased version once #6871 is merged of course.

sdaftuar committed:
Update replace-by-fee logic to use fee deltas

sdaftuar commented:
Rebased.

@petertodd This pull has now been updated to use fee deltas for the replace-by-fee calculation, and the replace-by-fee rpc test has been updated to exercise that logic.  (I cherry-picked #7137, so if that gets merged first I'll rebase this one, or I can close that pull if that's easier.)

petertodd commented on qa/rpc-tests/replace-by-fee.py:
```diff
@@ -63,8 +63,16 @@ def make_utxo(node, amount, confirmed=True, scriptPubKey=CScript([1])):
+                # the transactions we needed mined were mined, and
+                # don't worry about what is left.
+                break
+            mempool_size = new_size
```
FYI, I've found bugs before by erroring out if any txs are stuck; not sure this is a good idea.

sdaftuar replied:
Fair enough -- I ran into an issue where the really large transaction created in one of the tests was too big to mined, but I should be able to rework that test and make this check error out instead.

sdaftuar commented:
Rebased and changed rpc test to error out if the mempool ends up with stuck transactions.

sdaftuar commented:
Rebased now that #6898 is merged

sdaftuar committed:
Use fee deltas for determining mempool acceptance

sdaftuar committed:
Remove GetMinRelayFee

One test in AcceptToMemoryPool was to compare a transaction's fee
agains the value returned by GetMinRelayFee. This value was zero for
all small transactions.  For larger transactions (between
DEFAULT_BLOCK_PRIORITY_SIZE and MAX_STANDARD_TX_SIZE), this function
was preventing low fee transactions from ever being accepted.

With this function removed, we will now allow transactions in that range
with fees (including modifications via PrioritiseTransaction) below
the minRelayTxFee, provided that they have sufficient priority.

sdaftuar commented:
Updated so that now mempool acceptance is also determined by including a transaction's fee delta from PrioritiseTransaction.

In doing so, I realized that `GetMinRelayFee` is not very useful, so I scrapped it.  With the DEFAULT_BLOCK_PRIORITY_SIZE set to 0, this function was stopping all free transactions from making it in (unless they were prioritised using PrioritiseTransaction).  Scrapping it means that there is a behavior change from before, where now we will allow in transactions between 50kb and 100kb that are free as long as they have sufficient priority.  Given that when the mempool is full we don't allow free transactions at all, and given that we expire things after 3 days, I think this should be okay.

morcos commented:
utACK (modulo comment above)

sdaftuar commented:
Addressed @morcos' nit (and squashed into first commit).

laanwj commented:
utACK
